src/_pytest/cacheprovider.py

```diff
@@ -203,17 +205,35 @@ def set(self, key: str, value: object) -> None:
             with f:
                 f.write(data)
 
-    def _ensure_supporting_files(self) -> None:
-        """Create supporting files in the cache dir that are not really part of the cache."""
-        readme_path = self._cachedir / "README.md"
-        readme_path.write_text(README_CONTENT, encoding="UTF-8")
+    def _ensure_cache_dir_and_supporting_files(self) -> None:
+        """Create the cache dir and its supporting files."""
+        if self._cachedir.is_dir():
+            return
 
-        gitignore_path = self._cachedir.joinpath(".gitignore")
-        msg = "# Created by pytest automatically.\n*\n"
-        gitignore_path.write_text(msg, encoding="UTF-8")
+        files: Iterable[Tuple[str, Union[str, bytes]]] = (
+            ("README.md", README_CONTENT),
+            (".gitignore", "# Created by pytest automatically.\n*\n"),
+            ("CACHEDIR.TAG", CACHEDIR_TAG_CONTENT),
+        )
 
-        cachedir_tag_path = self._cachedir.joinpath("CACHEDIR.TAG")
-        cachedir_tag_path.write_bytes(CACHEDIR_TAG_CONTENT)
+        with tempfile.TemporaryDirectory() as d:
```

Comment:
This is not needed, but I think it is good taste for tools to use some prefix in case the directories are lingering for some reason: 

```suggestion
        with tempfile.TemporaryDirectory(prefix="pytest-cache-files-") as d:
```

Reply:
Also note this comment by @bluetech:

> create the dir in a TemporaryDirectory (in same dir as the cache dir so it's in the same filesystem)

Reply:
I don't think we should do that - we may end up with that directory lingering in the interrupt case. I don't know enough about how signal handlers work in python.

Reply:
i observe consistent rename failures on my own system when going from the tmpdir to the target


i maintain that if a copytree fallback is consistently needed, its over-complicated and should just write the files directly

Reply:
it should be noted that it wont fail in the testsuite if pytest itself as the cache dir targets for those would be on the same tmpfs